src/ugeneui/src/main_window/CreateDesktopShortcutTask.cpp

```diff
@@ -0,0 +1,193 @@
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
+ * MA 02110-1301, USA.
+ */
+
+#include "CreateDesktopShortcutTask.h"
+
+#if defined(Q_OS_LINUX)
+#include <QCoreApplication>
+#include <QDir>
+#include <QFile>
+#endif // Q_OS_LINUX
+
+#include <QMainWindow>
+#include <QMessageBox>
+#include <QNetworkReply>
+#include <QPushButton>
+
+#include <U2Core/AppContext.h>
+#include <U2Core/AppSettings.h>
+#include <U2Core/NetworkConfiguration.h>
+#include <U2Core/Settings.h>
+#include <U2Core/SyncHttp.h>
+#include <U2Core/U2SafePoints.h>
+
+#include <U2Gui/MainWindow.h>
+
+#if defined(Q_OS_WIN)
+#include <windows.h>
+#include <winnls.h>
+#include <shobjidl.h>
+#include <objbase.h>
+#include <objidl.h>
+#include <shlguid.h>
+#include <shlobj.h>
+#endif // Q_OS_WIN
+
+namespace U2 {
+
+CreateDesktopShortcutTask::CreateDesktopShortcutTask(bool startUp)
+    : Task(tr("Create desktop shortcut"), TaskFlag_NoRun) {
+    runOnStartup = startUp;
+    setVerboseLogMode(true);
+    startError = false;
+}
+
+void CreateDesktopShortcutTask::run() {
+    stateInfo.setDescription(tr("Create desktop shortcut running"));
+
+    stateInfo.setDescription(QString());
+}
+
+bool CreateDesktopShortcutTask::createDesktopShortcut() {
+#if defined(Q_OS_WIN)
+    HRESULT hres;
+    IShellLink *psl;
+
+    // Get a pointer to the IShellLink interface. It is assumed that CoInitialize
+    // has already been called.
+    hres = CoCreateInstance(CLSID_ShellLink, NULL, CLSCTX_INPROC_SERVER, IID_IShellLink, (LPVOID *)&psl);
+    if (SUCCEEDED(hres)) {
+        IPersistFile *ppf;
+
+        // Set the path to the shortcut target and add the description.
+        WCHAR path[MAX_PATH];
+        GetModuleFileNameW(NULL, path, MAX_PATH);
+        psl->SetPath(path);
+        psl->SetDescription(L"Unipro UGENE");
+
+        // Query IShellLink for the IPersistFile interface, used for saving the
+        // shortcut in persistent storage.
+        hres = psl->QueryInterface(IID_IPersistFile, (LPVOID *)&ppf);
+
+        if (SUCCEEDED(hres)) {
+            WCHAR wsz[MAX_PATH + 1];
+            CHAR pathLink[MAX_PATH + 1];
+
+            if (SHGetSpecialFolderPathA(HWND_DESKTOP, pathLink, CSIDL_DESKTOP, FALSE)) {
+                strncat(pathLink, "\\UGENE.lnk", __min(strlen("\\UGENE.lnk"), MAX_PATH - strlen(pathLink)));
+
+                // Ensure that the string is Unicode.
+                MultiByteToWideChar(CP_ACP, 0, pathLink, -1, wsz, MAX_PATH);
+
+                // Add code here to check return value from MultiByteWideChar
+                // for success.
+
+                // Save the link by calling IPersistFile::Save.
+                hres = ppf->Save(wsz, TRUE);
+                ppf->Release();
+            } else {
+                return false;

```

Comment:
Sometimes psl is released and sometimes is not?

Reply:
fixed